File: week5/py/split_array.py
# ///https://codeforces.com/edu/course/2/lesson/6 for explanations of bs problems
def isgood(mid, k, v):
    segments = 1
    currentsum = 0
    for i in v:
        currentsum += i
        if currentsum > mid:
            segments += 1
            currentsum = i
    return segments <= k

n, k = map(int, input().split())  # Read the long long n and k from input

v = list(map(int, input().split()))

# Initialize l and r
l = max(v)  # Initialize l to the maximum value in the list v
r = sum(v)  # Initialize r to the sum of all elements in the list v

while l < r:
    mid = (l + r) // 2
    if isgood(mid, k, v):
        r = mid
    else:
        l = mid + 1

print(r)  # Output the result in Python

File: week5/cpp/split_array.cpp
///https://codeforces.com/edu/course/2/lesson/6 for explanations of bs problems
#include <bits/stdc++.h>
#include <numeric>
#include <vector>
using namespace std;

bool isgood(long long mid, long long k, vector<int> &v) {
  long long segments = 1;
  long long currentsum = 0;
  for (auto i : v) {
    currentsum += i;
    if (currentsum > mid) {
      segments++;
      currentsum = i;
    }
  }
  return segments <= k;
}

int main() {
  long long n, k;
  cin >> n >> k;
  vector<int> v(n);
  long long l, r;
  l = 0;
  r = 0;
  for (int i = 0; i < n; i++) {
    cin >> v[i];
    l = l > v[i] ? l : v[i];
    r += v[i];
  }

  while (l < r) {
    long long mid = (l + r) / 2;
    if (isgood(mid, k, v))
      r = mid;
    else
      l = mid + 1;
  }
  cout << r << endl;

  return 0;
}
